### ISSUE TITLE
docker run monstache Unable to connect to MongoDB

### ISSUE THREAD

**Issue by codeTranMan** [rwynn/monstache#520]
Unable to connect to MongoDB using URL mongodb://localhost:27017: server selection error: server selection timeout, current topology: { Type: Unknown, Servers: [{ Addr: localhost:27017, Type: Unknown, Average RTT: 0, Last error: connection() error occured during connection handshake: dial tcp 127.0.0.1:27017: connect: connection refused }, ] }

**Comment by dirk39** [comment]
Hi, are you running MongoDB and Monstache in two different containers? If yes, you have to share the network between containers and contact the MongoDB container with his hostname.

**Comment by vipulg13** [comment]
Hi @dirk39, hi @rwynn, I am facing the same issue. I tried multiple options, suggested in other similar issues, however, none of them worked so far. In my case, the MongoDB and Monstache are running in two separate containers. As suggested, I have shared the network between the containers. However, when it comes to setting a hostname, I have tried multiple ways, but no success so far. Here are the settings:

---
1. the mongo uri in the config.toml file:
mongo-url = "mongodb://admin:password@localhost:27017/ir?connect=direct&authSource=admin&readPreference=primary"

Note: Since I have no replica set configured, I am setting "connect=direct". I have also tried <container name>, 127.0.0.1, "ea8e4af25ff3" instead of localhost.
---
2. the "docker inspect mongodb" gives following result for the config section:
Config": {
            "Hostname": "ea8e4af25ff3",
            "Domainname": "",
            "User": "",
            "AttachStdin": false,
            "AttachStdout": false,
            "AttachStderr": false,
            "ExposedPorts": {
                "27017/tcp": {}
            }
---
Kindly help me to resolve it. I am happy to share other configuration objects as well to scrutinize the issue.

**Comment by vipulg13** [comment]
Hi @dirk39, hi @rwynn, I am facing the same issue. I tried multiple options, suggested in other similar issues, however, none of them worked so far. In my case, the MongoDB and Monstache are running in two separate containers. As suggested, I have shared the network between the containers. However, when it comes to setting a hostname, I have tried <container name>, 127.0.0.1, "ea8e4af25ff3" instead of localhost, but no success so far. Also, since I have no replica set configured, I am setting "connect=direct".

---
1. the mongo uri in the config.toml file:
mongo-url = "mongodb://admin:****@localhost:27017/ir?connect=direct&authSource=admin&readPreference=primary"
---
2. the "docker inspect mongodb" gives following result for the config section:
Config": {
            "Hostname": "ea8e4af25ff3",
            "Domainname": "",
            "User": "",
            "AttachStdin": false,
            "AttachStdout": false,
            "AttachStderr": false,
            "ExposedPorts": {
                "27017/tcp": {}
            }
---
Kindly help me to resolve it. I am happy to share other configuration objects as well to scrutinize the issue.

**Comment by routel-nolan** [comment]
I have the same problem here. I tried everything I saw on the web but nothing worked.
![Screenshot 2022-05-03 at 16 46 33](https://user-images.githubusercontent.com/47741919/166476466-6e0525b5-572e-44a5-a63b-27043cf5ce3b.png)

**Comment by dirk39** [comment]
Hi @Zardows , if you are running monstache in a container, or using docker-compose, you cannot connect to mongo through localhost, but using the container name or the service name.

**Comment by routel-nolan** [comment]
Hi @dirk39 ty for your answer. So if I understood what you told me I have to replace the mongo-url by something else ?
![Screenshot 2022-05-03 at 17 19 47](https://user-images.githubusercontent.com/47741919/166483286-75ff407c-7ff4-4473-b691-aa2f3aea7d6e.png)

**Comment by dirk39** [comment]
@Zardows yes, you should replace the host with the host of es and MongoDB.

**Comment by routel-nolan** [comment]
@dirk39 but es and Mongo are on my localhost so it should work no ?

**Comment by dirk39** [comment]
are you running monstache in the host or inside the container?

**Comment by routel-nolan** [comment]
in the container

**Comment by dirk39** [comment]
so the network is not the same. I mean, the localhost of the container network is not the same as the host machine. You can have a look [here](https://stackoverflow.com/questions/31324981/how-to-access-host-port-from-docker-container) to allow communication between the container and the host machine. BTW, I suggest adding MongoDB and ES in docker leveraging docker-compose.

**Comment by routel-nolan** [comment]
I tried with the gateway and it still doesn't work. I know for the Mongo on the docker-compose but I can't because I have to create 26 docker-compose file to create so it will not be optimised

**Comment by dirk39** [comment]
what do you think to run monstache as an executable instead of using a container?

**Comment by routel-nolan** [comment]
I need it to be a container unfortunately
